src/Psl/Str/replace_ci.php

```diff
@@ -12,12 +14,16 @@
  * `$replacement` (case-insensitive).
  *
  * @pure
+ *
+ * @throws Psl\Exception\InvariantViolationException if $needle is not a valid UTF-8 string.
  */
 function replace_ci(string $haystack, string $needle, string $replacement, Encoding $encoding = Encoding::UTF_8): string
 {
     if ('' === $needle || null === search_ci($haystack, $needle, 0, $encoding)) {
         return $haystack;
     }
 
+    Psl\invariant(is_utf8($needle), 'Expected $needle to be a valid UTF-8 string.');
+
     return join(preg_split('{' . preg_quote($needle, '/') . '}iu', $haystack), $replacement);
```

Comment:
Nice, I can easily rewrite the PR to null checks, but this function will output a warning: https://3v4l.org/citBH

Should we keep there the utf8 invariant, or we would wrap the call to `preg_split` in some way?

Reply:
You might want to take a look at `Internal\suppress()` or `Internal\box()` - which are probably the best way to go.

But if I benchmark this, the fastest way is 'silencing' the error:

```php
#[Groups(['str'])]
final class Utf8Bench
{
    public function benchValidUtf8CheckThroughPreg(): void
    {
        if (null === preg_replace('/./u', '', "HELLO")) {
            throw new \RuntimeException('xxx');
        }
    }

    public function benchValidUtf8CheckThroughPregSuppressed(): void
    {
        if (null === Internal\suppress(static fn () => preg_replace('/./u', '', "HELLO"))) {
            throw new \RuntimeException('xxx');
        }
    }

    public function benchValidUtf8CheckThroughPregSilenced(): void
    {
        if (null === @preg_replace('/./u', '', "HELLO")) {
            throw new \RuntimeException('xxx');
        }
    }
}

```

```
\Psl\Tests\Benchmark\Str\Utf8Bench

    benchValidUtf8CheckThroughPreg..........I4 - Mo1.000μs (±0.00%)
    benchValidUtf8CheckThroughPregSuppresse.R2 I3 - Mo2.300μs (±1.72%)
    benchValidUtf8CheckThroughPregSilenced..R2 I4 - Mo0.800μs (±0.00%)

Subjects: 3, Assertions: 0, Failures: 0, Errors: 0
Utf8Bench
=========

Average iteration times by variant

2.3μs     │   █
2.0μs     │   █
1.7μs     │   █
1.4μs     │   █
1.2μs     │ ▄ █
0.9μs     │ █ █ ▇
0.6μs     │ █ █ █
0.3μs     │ █ █ █
          └───────
            1 2 3
<current>
+---------------------------------------------+---------+---------+--------+---------+
| subject                                     | memory  | mode    | rstdev | stdev   |
+---------------------------------------------+---------+---------+--------+---------+
| benchValidUtf8CheckThroughPreg ()           | 6.068mb | 1.000μs | ±0.00% | 0.000μs |
| benchValidUtf8CheckThroughPregSuppressed () | 6.068mb | 2.300μs | ±1.72% | 0.040μs |
| benchValidUtf8CheckThroughPregSilenced ()   | 6.068mb | 0.800μs | ±0.00% | 0.000μs |
+---------------------------------------------+---------+---------+--------+---------+
```

Not sure how strict @azjezz policy is on using silencing operators - since it will scream if `boolean xdebug.scream = true`

Reply:
we can go with `Internal\suppress()`, the thing that bothers me about `@` is as you said, it would be ignored in case `xdebug,scream` is set to `true`, but other than that, i have no problem using `@` really here.